src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs

```diff
@@ -234,17 +279,50 @@ private uint GetPlaceholderScope(BoundValuePlaceholderBase placeholder)
         public override BoundNode? Visit(BoundNode? node)
         {
 #if DEBUG
-            if (node is BoundValuePlaceholderBase placeholder
-                // CheckValEscapeOfObjectInitializer() does not use BoundObjectOrCollectionValuePlaceholder.
-                // CheckInterpolatedStringHandlerConversionEscape() does not use BoundInterpolatedStringHandlerPlaceholder.
-                && node is not (BoundObjectOrCollectionValuePlaceholder or BoundInterpolatedStringHandlerPlaceholder))
+            if (node is BoundExpression expr)
+            {
+                bool added = MarkVisited(expr);
+                if (!added)
+                {
+                    Debug.Assert(expr is BoundValuePlaceholderBase);
+                }
+            }
+            if (node is BoundValuePlaceholderBase placeholder)
             {
-                Debug.Assert(_placeholderScopes?.ContainsKey(placeholder) == true);
+                Debug.Assert(ContainsPlaceholderScope(placeholder));
             }
 #endif
             return base.Visit(node);
         }
 
+#if DEBUG
+        private void CheckVisited(BoundExpression expr)
+        {
+            if (expr is BoundValuePlaceholderBase placeholder
+                && placeholder is not BoundCapturedReceiverPlaceholder)
```

Comment:
A comment on BoundCapturedReceiverPlaceholder would be useful (why we have this exception).